src/style/css_syntax/values.rs

```diff
@@ -0,0 +1,333 @@
+        let x = T::parse(input, taffy)?;
+        let y = if input.is_exhausted() { x.clone() } else { T::parse(input, taffy)? };
+        Ok(Point { x, y })
+    }
+
+    fn serialize(&self, dest: &mut String, taffy: &Taffy) {
+        self.x.serialize(dest, taffy);
+        if self.x != self.y {
+            dest.push(' ');
+            self.y.serialize(dest, taffy);
+        }
+    }
+}
+
+/// Assumes `<top> (<right> (<bottom> <left>?)?)?` syntax like `margin` etc.
+///
+/// <https://w3c.github.io/csswg-drafts/css-box-4/#propdef-margin>
+impl<T: CssValue + PartialEq + Clone> CssValue for Rect<T> {
+    fn parse<'i, 't>(input: &mut Parser<'i, 't>, taffy: &Taffy) -> Result<Self, ParseError<'i>> {
+        let top = T::parse(input, taffy)?;
+        let right = if input.is_exhausted() { top.clone() } else { T::parse(input, taffy)? };
+        let bottom = if input.is_exhausted() { top.clone() } else { T::parse(input, taffy)? };
+        let left = if input.is_exhausted() { right.clone() } else { T::parse(input, taffy)? };
+        Ok(Rect { top, right, bottom, left })
+    }
+
+    fn serialize(&self, dest: &mut String, taffy: &Taffy) {
+        let right_is_needed = self.right != self.top;
+        let bottom_is_needed = self.bottom != self.top;
+        let left_is_needed = self.left != self.right;
+        self.top.serialize(dest, taffy);
+        if right_is_needed || bottom_is_needed || left_is_needed {
+            dest.push(' ');
+            self.right.serialize(dest, taffy);
+            if bottom_is_needed || left_is_needed {
+                dest.push(' ');
+                self.bottom.serialize(dest, taffy);
+                if left_is_needed {
+                    dest.push(' ');
+                    self.left.serialize(dest, taffy);
+                }
+            }
+        }
+    }
+}
+
+impl CssValue for LengthPercentage {
+    fn parse<'i, 't>(input: &mut Parser<'i, 't>, taffy: &Taffy) -> Result<Self, ParseError<'i>> {
+        let token = input.next()?.clone();
+        match &token {
+            Token::Dimension { value, unit, .. } => {
+                // We do not support relative length units are not supported
+                // Absolute units all have a fix ratio to each other:
+                // https://drafts.csswg.org/css-values/#absolute-lengths
+                let units_per_inch = match_ignore_ascii_case! { &*unit,
+                    "px" => 96.,
+                    "pt" => 72., // point
+                    "pc" => 6., // pica
+                    "in" => 1.,
+                    "cm" => 2.54,
+                    "mm" => 25.4,
+                    "q" =>  25.4 * 4., // quarter millimeter
+                    _ => return Err(input.new_unexpected_token_error(token))
+                };
+                let css_inches = value / units_per_inch;
+                let css_pxs = css_inches * 96.;
+                let taffy_units = css_pxs * taffy.config.pixel_ratio;
```

Comment:
I think it might make more sense to have `Dimension::Length` store logical pixels. That would simplify this code (it wouldn't have to care about `pixel_ratio`: it can just naively pass `px` values through to the `Length` variant), and would also match how I would expect users of Taffy's underlying Rust API to specify styles.

I think the `pixel_ratio` config option on the `Taffy` struct is really good idea, and actually *enables* this parsing code to ignore pixel ratio, becuase it means that Taffy can operate on a "logical pixels in, physical pixels out basis" (or perhaps even "unrounded logical pixels AND rounded physical pixels out").

Reply:
So `Dimension::Length` would be in CSS `px` and values in `Layout` would *not* be in the same unit anymore but differ by `pixel_ratio`? That could work, but I feel the latter half should be a separate PR. But yeah it’d be nice not to need to pass `&Taffy` around in all parsing and serialization functions.

Reply:
https://github.com/DioxusLabs/taffy/pull/478 attempts to implement your proposal, which would replace conversion during parsing and serialization.